CountingBloomFilter.hpp

```diff
@@ -271,61 +252,76 @@ template<typename T>
 void
 CountingBloomFilter<T>::readFilter(const std::string& path)
 {
-	FILE* file;
-	if ((file = fopen(path.c_str(), "rb")) == nullptr) {
+	std::ifstream file(path);
+	if (!file.is_open()) {
 		std::cerr << "ERROR: Failed to open file: " << path << "\n";
 		exit(EXIT_FAILURE);
 	}
 	readHeader(file);
-	struct stat buf;
-	if (fstat(fileno(file), &buf) != 0) {
-		std::cerr << "ERROR: Failed to open file: " << path << "\n";
-		exit(EXIT_FAILURE);
-	}
-	size_t arraySizeOnDisk = buf.st_size - sizeof(struct FileHeader);
-	if (arraySizeOnDisk != m_sizeInBytes) {
-		std::cerr << "ERROR: File size of " << path << " (" << arraySizeOnDisk << " bytes), "
-		          << "does not match size read from its header (" << m_sizeInBytes << " bytes).\n";
-		exit(EXIT_FAILURE);
-	}
-
-	size_t nread = fread(m_filter, arraySizeOnDisk, 1, file);
-	if (nread != 1 && fclose(file) != 0) {
+	char* filter;
+	filter = new char[m_sizeInBytes];
+	file.read(filter, m_sizeInBytes);
+	m_filter = reinterpret_cast<T*>(filter);
+	if (!file) {
 		std::cerr << "ERROR: The byte array could not be read from the file: " << path << "\n";
 		exit(EXIT_FAILURE);
 	}
+	file.close();
 }
 
 template<typename T>
 void
-CountingBloomFilter<T>::readHeader(FILE* file)
+CountingBloomFilter<T>::readHeader(std::istream& file)
 {
-	FileHeader header;
-	if (fread(&header, sizeof(struct FileHeader), 1, file) != 1) {
-		std::cerr << "Failed to read header\n";
-		exit(EXIT_FAILURE);
-	}
-	if (header.hlen != sizeof(FileHeader)) {
-		std::cerr << "Bloom Filter header length: " << header.hlen
-		          << " does not match expected length: " << sizeof(FileHeader)
-		          << " (likely version mismatch)" << std::endl;
-		exit(EXIT_FAILURE);
-	}
-	if (memcmp(header.magic, MAGIC_HEADER_STRING, MAGIC_LENGTH) != 0) {
-		std::cerr << "Bloom Filter type does not match" << std::endl;
+	std::string magic_header(MAGIC_HEADER_STRING);
+	(magic_header.insert(0, "[")).append("]");
+	std::string line;
+	std::getline(file, line);
+	if (line.compare(magic_header) != 0) {
+		std::cerr << "ERROR: magic string does not match (likely version mismatch) \n";
+		std::cerr << "Your magic string:                " << line << "\n";
+		std::cerr << "CountingBloomFilter magic string: " << magic_header << std::endl;
 		exit(EXIT_FAILURE);
 	}
-	if (header.version != BloomFilter_VERSION) {
-		std::cerr << "Bloom Filter version does not match: " << header.version
-		          << " expected: " << BloomFilter_VERSION << std::endl;
-		exit(EXIT_FAILURE);
+
+	/* Read bloom filter line by line until it sees "[HeaderEnd]"
+	   which is used to mark the end of the header section and
+	   assigns the header to a char array*/
+	std::string headerEnd = "[HeaderEnd]";
+	char* toml_buffer = new char[0];
```

Comment:
```suggestion
	std::vector<char> toml_buffer;
```
Use a `std::vector` instead of a C array. Always prefers a vector unless you have a good reason to use an array.

Reply:
so there is a reason which is that ```read(char_type* __s, streamsize __n);```` requires char type array. I can do
```
std::vector<char> toml_buffer;
toml_buffer.resize(currPos);
file.read (&toml_buffer[0], currPos);```
With regards to ```std::vector<uint8_t> filter([m_sizeInBytes]);```,  I can do ```
std::vector<char> filter([m_sizeInBytes]);
file.read (&filter[0],m_sizeInBytes );
m_filter = reinterpret_cast<T*>(&filter[0]);
``` but I can't ```std::vector<uint8_t> filter([m_sizeInBytes]);file.read (&filter[0],m_sizeInBytes );```

Reply:
`m_filter` should be a `std::vector`. Cast the argument to `read`.

Reply:
you mean something like this? 

```
std::vector<uint8_t> m_filter([m_sizeInBytes]);
file.read (reinterpret_cast<char*>(&filter[0]), m_sizeInBytes );

```

Reply:
Yep, that looks about right. `m_filter([m_sizeInBytes])` should I think be `m_filter(m_sizeInBytes)`.

Reply:
`reinterpret_cast` should I think be `static_cast`.